### ISSUE TITLE
Incorrect values when converting DATETIME to TIMESTAMP on non-UTC server

### ISSUE THREAD

**Issue by mignaulo** [github/gh-ost#247]
We are seeing incorrect date values when converting a DATETIME column to TIMESTAMP on a server which is not in UTC. The dates are off by the number of hours between PDT (the server timezone) and UTC. 

To explain why we are doing this, we are attempting to repair the bad decisions we made when we didn't have our MySQL servers in UTC. We are converting our DATETIME columns to TIMESTAMP which will allow us to easily change the database time zone. 

Rows that were copied in batches are correct. I assume this is performed using the default server timezone, the tool does not try to override that.

Rows that are inserted from the binlog are incorrect. I assume the binlog is inherently in the timezone of the server (in this case PDT), while inserts are specifically made with the time_zone set to UTC (https://github.com/github/gh-ost/blob/master/go/logic/applier.go#L875). 
We therefore end up with some rows 7 hours behind what we expect. 

I will need to do further testing, but binlog update statements also do not seem to apply. I am not certain about this, but I assume the UPDATE statement it generates might include every old column values in the WHERE clause? If that is the case, the timestamp in the table would not match, since changing the time_zone changes the table date to UTC while the date from the binlog is still in PDT.

I believe this is related to #161 or #175, and fixing that particular issue probably caused the one I am currently seeing. 

I'm unsure what would be the best way to fix this problem. Removing the session time_zone would certainly fix my problem, but would bring back the problems from the issues mentioned above. Being able to override the timezone this is set to would also help but it's not the cleanest of solutions.

Happy to provide additional information and run tests. 

I appreciate your help here, thanks!

gh-ost: 1.0.20

**Comment by shlomi-noach** [comment]
https://github.com/github/gh-ost/pull/248 adds a test for the above use case. Right now, that test fails.

**Comment by shlomi-noach** [comment]
> To explain why we are doing this, we are attempting to repair the bad decisions we made when we didn't have our MySQL servers in UTC. We are converting our DATETIME columns to TIMESTAMP which will allow us to easily change the database time zone. 

Unrelated to the bug, this is an interesting note. I'm in fact more inclined to get rid of `timestamp` as much as possible. One option is to use `datetime`, another is to store time as `int` in `unix_timestamp` format (ie seconds since leap time).

Reason is the insane conversions when connecting with utc/non-utc settings; the outputting of time as string, which is then difficult to manipulate.

Food for thought.

**Comment by shlomi-noach** [comment]
As per the bug: the bug exists, because same time values are copied in two different ways (by row-copy and by binlog event apply) into two different values.

However, I'm actually unsure which is the correct one. Reason says that the rowcopy are the correct ones, because that's the direct result of a `insert into ... select` query directly on the server.
However that may actually be incorrect, and largely depends on what the "assumed" timestamp was for your `datetime` values. Is it safe to assume those values were in your local timezone? (I'm writing "your" but this is a question for the general case).

Maybe you stored your `datetime` in `UTC`, intentionally? I'm unsure there's a definitive answer here, and am inclined to just go with whatever behavior `mysql` chose to use.

I'm also wondering whether there should be any user defined configuration to this.

cc @dveeden @ggunson who may have more insight.

**Comment by shlomi-noach** [comment]
#248 's https://github.com/github/gh-ost/pull/248/commits/33ebc2d177cb93c49c7bd855bef1fb92c22c022e seems to fix the issue. Nonetheless I'd like to have that discussion.

**Comment by ggunson** [comment]
I would think the correct temporal value is whatever a straight `ALTER TABLE` would do, and from my testing it assumes the session timezone (so, the master timezone unless specified otherwise) of the session that runs the alter for how to interpret the datetime:

```
#160928 15:07:30 server id 1  end_log_pos 886 CRC32 0xe1c70b04 	Query	thread_id=37	exec_time=0	error_code=0
SET TIMESTAMP=1475075250/*!*/;
SET @@session.sql_mode=1437073440/*!*/;
SET @@session.time_zone='-07:00'/*!*/;
alter table timefoo3 modify t2 timestamp default current_timestamp
/*!*/;
```

(This is from the replica's binlog; in this case I used a non-UTC on the master and UTC on the replica).

**Comment by mignaulo** [comment]
Interesting thoughts regarding `timestamp` columns and just using `int` - I appreciate it.

Regarding #248, I tested it and did not have any luck (same incorrect behavior). I did find that the modified code does not seem to run in my case. The field type is always MYSQL_TYPE_DATETIME, not MYSQL_TYPE_DATETIME2. I'm not quite sure what is the difference (Antelope vs Barracuda? fractional time?), this server is running MySQL 5.6. Any ideas?

The time string that comes out of the MYSQL_TYPE_DATETIME parsing is effectively in non-UTC.

Thanks!

**Comment by shlomi-noach** [comment]
@mignaulo strange. I'm also testing on `5.6`. Can you please paste your `CREATE TABLE`? Also, the existing behavior for `MYSQL_TYPE_DATETIME` is to modify tz to `UTC()`, which is the correct handling.

**Comment by shlomi-noach** [comment]
@mignaulo I'm also curious about the `ENGINE` and `ROW_FORMAT` for your table.

**Comment by shlomi-noach** [comment]
According to [documentation](https://dev.mysql.com/doc/refman/5.6/en/c-api-data-structures.html):

> The MYSQL_TYPE_TIME2, MYSQL_TYPE_DATETIME2, and MYSQL_TYPE_TIMESTAMP2) type codes are used only on the server side. Clients see the MYSQL_TYPE_TIME, MYSQL_TYPE_DATETIME, and MYSQL_TYPE_TIMESTAMP codes. 

which makes my puzzled.

**Comment by shlomi-noach** [comment]
The test results are inconsistent; I'm able to make it fail when using `ROW_FORMAT=DYNAMIC` -- but not always. I'm very puzzled at this because the test is mostly deterministic (other than using `now()`).

**Comment by shlomi-noach** [comment]
I'm yet unable to generate a table that has a `MYSQL_TYPE_DATETIME`

**Comment by mignaulo** [comment]
@shlomi-noach, the `DATETIME` vs `DATETIME2` thing is almost unexplainable. When I create a new table based on an existing table ie `CREATE TABLE x LIKE y`, the table has a `MYSQL_TYPE_DATETIME`. When I create the very same table using a full `CREATE TABLE` statement that defines the rows, it has a `MYSQL_TYPE_DATETIME2` type.  These tables were created a long time ago on previous versions of MySQL, I'm not sure I can help you reproduce that unfortunately.

To reduce to a case we can both reproduce, I'm now using the following table which has a `MYSQL_TYPE_DATETIME2` column.
```
CREATE TABLE `_test_test` (
  `id` int(11) unsigned NOT NULL AUTO_INCREMENT,
  `date` datetime NOT NULL,
  `ip_address` int(11) unsigned NOT NULL,
  `val1` int(11) unsigned NOT NULL DEFAULT '0',
  `val2` int(11) unsigned NOT NULL DEFAULT '0',
  `val3` smallint(6) unsigned NOT NULL DEFAULT '0',
  `latitude` decimal(8,4) NOT NULL,
  `longitude` decimal(8,4) NOT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=1798910615 DEFAULT CHARSET=utf8 ROW_FORMAT=DYNAMIC
```

I start a test migration and run: 
`insert into _test_test values (1798910616, "2016-09-29 11:15:23", 533817856, 345454,23423, 34, 45.5444, 34.3333);`

Adding some debug output to the row_event code, both the old and the new code produce the same time strings:

datetime2 new 2016-09-29 11:15:23
datetime2 old 2016-09-29 11:15:23

(For my case at least, the new code leads to the same time zone discrepancy)

Then looking at the tables: 
```
mysql> select * from _test_test where id=1798910616;
+------------+---------------------+------------+--------+-------+------+----------+-----------+
| id         | date                | ip_address | val1   | val2  | val3 | latitude | longitude |
+------------+---------------------+------------+--------+-------+------+----------+-----------+
| 1798910616 | 2016-09-29 11:15:23 |  533817856 | 345454 | 23423 |   34 |  45.5444 |   34.3333 |
+------------+---------------------+------------+--------+-------+------+----------+-----------+
1 row in set (0.00 sec)

mysql> select * from __test_test_gho where id=1798910616;
+------------+---------------------+------------+--------+-------+------+----------+-----------+
| id         | date                | ip_address | val1   | val2  | val3 | latitude | longitude |
+------------+---------------------+------------+--------+-------+------+----------+-----------+
| 1798910616 | 2016-09-29 04:15:23 |  533817856 | 345454 | 23423 |   34 |  45.5444 |   34.3333 |
+------------+---------------------+------------+--------+-------+------+----------+-----------+
1 row in set (0.00 sec)
```

**Comment by shlomi-noach** [comment]
@MarkLeith could we possibly pick your brain re: `MYSQL_TYPE_DATETIME` and `MYSQL_TYPE_DATETIME2`? See above discussion

**Comment by shlomi-noach** [comment]
@mignaulo thank you for reducing to common grounds -- I'm continuing to pursue this.

**Comment by shlomi-noach** [comment]
@mignaulo, can you please pull & test again #248 ? I've identified the problem in the `time_zone` setting for the row-copy itself, and tests now pass consistently.

**Comment by mignaulo** [comment]
Hi @shlomi-noach,

I tested on the table above. Here is the comparison between the two tables:
```
mysql> select * from _test_test order by id desc limit 10;
+------------+---------------------+------------+-------+-------+------+----------+-----------+
| id         | date                | ip_address | val1  | val2  | val3 | latitude | longitude |
+------------+---------------------+------------+-------+-------+------+----------+-----------+
| 1810613255 | 2016-10-08 11:44:42 | 1610295655 | 18707 |  1004 |   22 |  43.5333 |   13.6167 |
| 1810613254 | 2016-10-08 11:44:42 |   39795770 | 44409 | 39085 |    2 |  45.1446 |   38.9942 |
| 1810613253 | 2016-10-08 11:44:42 |  823120780 |  4556 |  4145 |  100 |  20.0000 |   77.0000 |
| 1810613252 | 2016-10-08 11:44:42 | 1135993743 | 28699 |  6113 |   17 |  36.3417 | -119.2553 |
| 1810613251 | 2016-10-08 11:44:42 | 2535864572 | 13157 |  2411 |   95 |  43.1479 |   12.1097 |
| 1800147398 | 2016-09-30 09:54:36 | 3139196250 |  2381 |   468 |  236 |  -4.2264 |  -44.7925 |
| 1800147397 | 2016-09-30 09:54:36 | 3057679431 |  8000 |   718 |   19 |  28.6667 |   77.2167 |
| 1800147396 | 2016-09-30 09:54:36 | 1388824619 |  9286 |  1435 |   27 |  40.1812 |   44.5158 |
| 1800147395 | 2016-09-30 09:54:36 | 1571398290 |  9939 |   857 |   28 |  24.6537 |   46.7152 |
| 1800147394 | 2016-09-30 09:54:36 | 3152875218 | 20598 | 24174 |   28 |  19.3733 |  -99.1631 |
+------------+---------------------+------------+-------+-------+------+----------+-----------+
10 rows in set (0.00 sec)

mysql> select * from __test_test_gho order by id desc limit 10;
+------------+---------------------+------------+-------+-------+------+----------+-----------+
| id         | date                | ip_address | val1  | val2  | val3 | latitude | longitude |
+------------+---------------------+------------+-------+-------+------+----------+-----------+
| 1810613255 | 2016-10-08 04:44:42 | 1610295655 | 18707 |  1004 |   22 |  43.5333 |   13.6167 |
| 1810613254 | 2016-10-08 04:44:42 |   39795770 | 44409 | 39085 |    2 |  45.1446 |   38.9942 |
| 1810613253 | 2016-10-08 04:44:42 |  823120780 |  4556 |  4145 |  100 |  20.0000 |   77.0000 |
| 1810613252 | 2016-10-08 04:44:42 | 1135993743 | 28699 |  6113 |   17 |  36.3417 | -119.2553 |
| 1810613251 | 2016-10-08 04:44:42 | 2535864572 | 13157 |  2411 |   95 |  43.1479 |   12.1097 |
| 1800147398 | 2016-09-30 02:54:36 | 3139196250 |  2381 |   468 |  236 |  -4.2264 |  -44.7925 |
| 1800147397 | 2016-09-30 02:54:36 | 3057679431 |  8000 |   718 |   19 |  28.6667 |   77.2167 |
| 1800147396 | 2016-09-30 02:54:36 | 1388824619 |  9286 |  1435 |   27 |  40.1812 |   44.5158 |
| 1800147395 | 2016-09-30 02:54:36 | 1571398290 |  9939 |   857 |   28 |  24.6537 |   46.7152 |
| 1800147394 | 2016-09-30 02:54:36 | 3152875218 | 20598 | 24174 |   28 |  19.3733 |  -99.1631 |
+------------+---------------------+------------+-------+-------+------+----------+-----------+
```

The top five rows in each table were inserted into the original table during the copy (applied to new table via binlog). The bottom five rows were present in the original table during the copy. 

The row copy is now consistent with the binlog, but I'd argue that we are further away from what I'd expect to happen. The discrepancy between my results and your results likely comes from the timezone used. Both client and server are PST here, and the connection happens over PST. The data in the original table is considered to be PST. The tool assumes that the data is UTC. If I change the session time_zone to UTC to run the above queries, I do receive the same output between both tables.

If I may, I've been successfully running a couple of alter on the same servers I am testing this and my "patch" was to remove the SESSION time_zone='+0:00' override in `go/logic/applier.go`, basically reverting #175. From my point of view, the problem comes from trying to convert time values outside of MySQL. #175 fixes the problem for timestamp => timestamp alters, but breaks datetime => timestamp. I feel like modifying decodeTimestamp2 in row_event to return something like `FROM_UNIXTIME(timestamp)` instead of trying to convert the binlog into a date string to insert/update would take care of the original problem from #175 while leaving datetime => timestamp conversions working. 

Unfortunately, this is my first time working with go, so my knowledge is relatively limited and I have not been able to run the test suite locally. I can't tell if that would truly fix the original issues from #175 or if it would pass the testsuite, but if my understanding of the code is correct, it should.

Thanks for the hard work trying to deal with this issue. All your help here is definitely appreciated.

**Comment by shlomi-noach** [comment]
@mignaulo your detailed response is very helpful to debugging of this problem.
Getting to have consistency between row-copy and event-handling is a major concern; now we have a knob that controls the timezone-shift, that we can apply to both copies and remain consistent.

My argument right now is that there is no way to _know_ in advance the timezone assumed in your `DATETIME` columns. For you it's PDT, for others, still working within PDT timezone -- it would be UTC.

One easy way out of this would be to add a command line flag, which could take the form of:

- `gh-ost --time_zone='+00:00'`
- `gh-ost --time_zone='system'`

With this flag I will conveniently kick the ball over to your field, and put responsibility in your hands.
Otherwise I'm trying to see if there's a holistic solution that would apply to all. #175 came into life for a reason; can we reason both that one and yours into a single solution?

**Comment by jonahberquist** [comment]
@mignaulo can you check if the binary log events for your table are prefixed with `SET @@session.time_zone` statements, and if so, can you post them here?

I'm inclined to agree with @ggunson that whatever time_zone the straight alter would do is the right one (https://github.com/github/gh-ost/issues/247#issuecomment-250202546). By the same logic, whatever time_zone the statement is in should be the right one to use in the applier.

**Comment by shlomi-noach** [comment]
> whatever time_zone the straight alter would do is the right one (#247 (comment)). By the same logic, whatever time_zone the statement is in should be the right one to use in the applier.

Unfortunately that was the very original behavior and it presents errors in other tests; see original bug report in https://github.com/github/gh-ost/issues/161

**Comment by shlomi-noach** [comment]
@mignaulo will you kindly test https://github.com/github/gh-ost/pull/265 ? In this PR I'm trying out the unifying theorem of the universe, and more specifically the reverting of much of TZ related code, retracting `go-mysql` changes, but taking server `time_zone` as the basis for all conversions.

Please let me know if https://github.com/github/gh-ost/pull/265 works.

Noteworthy is that master & replica need to be in same timezone for this to work correctly.

**Comment by jonahberquist** [comment]
> Unfortunately that was the very original behavior and it presents errors in other tests; see original bug report in #161

I don't see this addressed in that issue. What I'm saying is that there can be time_zone information in the binary logs that would indicate what time_zone the statement is in.

As an example from a host in PDT, I created a table `time_zone` on my local machine with the following definition:
```
CREATE TABLE `time_zone` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `dt` datetime NOT NULL,
  `ts` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=7 DEFAULT CHARSET=utf8
```
I populated every row with:
```
insert into time_zone (dt, ts) values (now(), now());
```
Before some of my inserts, I changed the session time_zone. Check out these binary log entries:
```
# at 1151
#161010 16:26:16 server id 1  end_log_pos 1231 CRC32 0x677fcf06        	Query  	thread_id=1    	exec_time=0    	error_code=0
SET TIMESTAMP=1476141976/*!*/;
BEGIN
/*!*/;
# at 1231
#161010 16:26:16 server id 1  end_log_pos 1287 CRC32 0x2ceef466        	Table_map: `test`.`time_zone` mapped to number 77
# at 1287
#161010 16:26:16 server id 1  end_log_pos 1336 CRC32 0x635110c0        	Write_rows: table id 77 flags: STMT_END_F

BINLOG '
mCP8VxMBAAAAOAAAAAcFAAAAAE0AAAAAAAEABHRlc3QACXRpbWVfem9uZQADAxIRAgAAAGb07iw=
mCP8Vx4BAAAAMQAAADgFAAAAAE0AAAAAAAEAAgAD//gEAAAAmZqVBpBX/COYwBBRYw==
'/*!*/;
### INSERT INTO `test`.`time_zone`
### SET
###   @1=4
###   @2='2016-10-10 16:26:16'
###   @3=1476141976
# at 1336
#161010 16:26:16 server id 1  end_log_pos 1367 CRC32 0x46c329f8        	Xid = 39
COMMIT/*!*/;
# at 1367
#161010 16:26:32 server id 1  end_log_pos 1447 CRC32 0x743e71d5        	Query  	thread_id=1    	exec_time=0    	error_code=0
SET TIMESTAMP=1476141992/*!*/;
SET @@session.time_zone='+00:00'/*!*/;
BEGIN
/*!*/;
# at 1447
#161010 16:26:32 server id 1  end_log_pos 1503 CRC32 0xf012aeed        	Table_map: `test`.`time_zone` mapped to number 77
# at 1503
#161010 16:26:32 server id 1  end_log_pos 1552 CRC32 0xb2c67989        	Write_rows: table id 77 flags: STMT_END_F

BINLOG '
qCP8VxMBAAAAOAAAAN8FAAAAAE0AAAAAAAEABHRlc3QACXRpbWVfem9uZQADAxIRAgAAAO2uEvA=
qCP8Vx4BAAAAMQAAABAGAAAAAE0AAAAAAAEAAgAD//gFAAAAmZqVdqBX/COoiXnGsg==
'/*!*/;
### INSERT INTO `test`.`time_zone`
### SET
###   @1=5
###   @2='2016-10-10 23:26:32'
###   @3=1476141992
# at 1552
#161010 16:26:32 server id 1  end_log_pos 1583 CRC32 0x5ba7006b        	Xid = 42
COMMIT/*!*/;
# at 1583
#161010 16:26:40 server id 1  end_log_pos 1663 CRC32 0xd56bf409        	Query  	thread_id=1    	exec_time=0    	error_code=0
SET TIMESTAMP=1476142000/*!*/;
SET @@session.time_zone='SYSTEM'/*!*/;
BEGIN
/*!*/;
# at 1663
#161010 16:26:40 server id 1  end_log_pos 1719 CRC32 0xdf69280a        	Table_map: `test`.`time_zone` mapped to number 77
# at 1719
#161010 16:26:40 server id 1  end_log_pos 1768 CRC32 0x5949727b        	Write_rows: table id 77 flags: STMT_END_F

BINLOG '
sCP8VxMBAAAAOAAAALcGAAAAAE0AAAAAAAEABHRlc3QACXRpbWVfem9uZQADAxIRAgAAAAooad8=
sCP8Vx4BAAAAMQAAAOgGAAAAAE0AAAAAAAEAAgAD//gGAAAAmZqVBqhX/COwe3JJWQ==
'/*!*/;
### INSERT INTO `test`.`time_zone`
### SET
###   @1=6
###   @2='2016-10-10 16:26:40'
###   @3=1476142000
# at 1768
#161010 16:26:40 server id 1  end_log_pos 1799 CRC32 0x2295cb58        	Xid = 44
COMMIT/*!*/;
```

Note especially how the value in the datetime column changes time_zone when there has been a `SET @@session.time_zone='+00:00'/*!*/;` statement.

**Comment by mignaulo** [comment]
@shlomi-noach I think we are close to a winner! 

For the first attempt, I let gh-ost determine the MySQL timezone, it used SYSTEM which matched to Pacific. To be more thorough, I also added a timestamp column to the test table I was using. I can confirm both date columns were correct for the row copy and the binlog apply! 

> My argument right now is that there is no way to know in advance the timezone assumed in your DATETIME columns. For you it's PDT, for others, still working within PDT timezone -- it would be UTC.
That is definitely true. I tested the alter with the default (SYSTEM), America/Los_Angeles and America/New_York. 
Both SYSTEM and America/Los_Angeles were identical and matched my expectations.
I tried the alter with a different time zone (America/New_York), then changed my session time zone to match and both tables were again identical.

One gotcha I noticed in the code now: the binlog timestamp conversion is now done using local time - this is the local time of the host gh-ost is running on. If the client time zone differs from the server time zone (or the `--time-zone` value I assume), you will have problems with your TIMESTAMP columns created/updated by the binlog apply. Trying --time-zone SYSTEM (Pacific) with `TZ='America/New_York' ./gh-ost-265 ...` The timestamp is transformed into a string using EDT, but then written as PDT resulting in a timestamp column off by 3 hours. 

I'm not sure if that's just something to make clear in the documentation or if a code change could handle that case.


@jonahberquist The binlog events were not preceded by `SET @@session.time_zone` events normally as we don't change the time zone. When I tried the alter with "America/Los_Angeles" (technically the same timezone as SYSTEM, but still a different value) as the --time-zone and did end up creating entries that would switch the time_zone back and forth between SYSTEM (normal traffic) and America/Los_Angeles (alter)

**Comment by shlomi-noach** [comment]
@jonahberquist: ack: will see if I can parse the session `time_zone` from binary logs, if it appears.

@mignaulo 

> One gotcha I noticed in the code now: the binlog timestamp conversion is now done using local time - this is the local time of the host gh-ost is running on.

It's even slightly more complicated than that, because the master & replica themselves could be in different timezones.
Getting the timezone from the replica and setting it in the `go` app is somewhat complex. What would a `SYSTEM` time_zone mean on the `go` app? `SYSTEM` is like `Local`, but as you suggest, the `gh-ost` host could be on a completely different timezone itself. But then again, it gets no insight as for the time_zone on the server.

One thing I will do is to read master & replica timezones separately and apply them differntly in `applier.go`. Still need to consider app time_zone.

**Comment by shlomi-noach** [comment]
@mignaulo I'm going to pursue the unifying theorem by looking at master-replica-app each on their own different timezone. I'll open a new PR for that.

Having said that, modifying your `DATETIME` into `TIMESTAMP` on a setup where master and replica are in different timezones sounds unhealthy to me; making it so when the `gh-ost` timezone is different sounds the wrong thing to do -- particularly in light of your original quest.

Right now, the safest suggestion I can provide is, make master, replica, gh-ost on same timezone.
If your master uses RBR, you can migrate directly on master.

**Comment by shlomi-noach** [comment]
OK, some deeper investigation and I believe I fully understand the problem now.

@jonahberquist the binary logs present time in UTC (by default). While row-copy is trivial (`INSERT...SELECT`), applying the binlog data requires `gh-ost` to `set time_zone='+00:00'`.

For most `TIMESTAMP` related cases this is correct. However, the particular case of @mignaulo is different: he changes a `DATETIME` column into `TIMESTAMP`. A `DATETIME` is essentially a string, and the binary logs contain the very same string representation as provided by master.

Now, when trying to write this string into the new `TIMESTAMP` column in the _ghost_ table, conversion happens. For this particular case, we need to shove the value into the migrated server using the migrated server's `time_zone`. No matter how I look at this, this is a special behaviour that cannot be unified by a global behavior.

What concerns me right now is _what other weird cases there are_. I will look into modifying `TIMESTAMP` into `DATETIME` as well (likely to suffer same problem, opposite direction)

Looking further.

**Comment by shlomi-noach** [comment]
Moreover, issuing `set time_zone` is not granular enough, because we can have both `timestamp` and `datetime` columns in the same table, and they will need different approaches.

**Comment by shlomi-noach** [comment]
Another note, it is difficult to do the time conversion on `gh-ost`'s side. The MySQL time zone names are not identical to those respected by `go`. As example, `go` supports `US/Pacific` but not necessarily `-09:00`, which could be the TZ defined in MySQL.

**Comment by shlomi-noach** [comment]
A way through may be via:

```
$ ./master -e "select @@time_zone, now(), convert_tz(now(),@@time_zone,'+00:00') as utc, timestampdiff(hour, now(), convert_tz(now(),@@time_zone,'+00:00')) as diff"
+-------------+---------------------+---------------------+------+
| @@time_zone | now()               | utc                 | diff |
+-------------+---------------------+---------------------+------+
| -04:00      | 2016-10-12 11:15:12 | 2016-10-12 15:15:12 |    4 |
+-------------+---------------------+---------------------+------+
$ ./replica -e "select @@time_zone, now(), convert_tz(now(),@@time_zone,'+00:00') as utc, timestampdiff(hour, now(), convert_tz(now(),@@time_zone,'+00:00')) as diff"
+-------------+---------------------+---------------------+------+
| @@time_zone | now()               | utc                 | diff |
+-------------+---------------------+---------------------+------+
| -07:00      | 2016-10-12 08:15:18 | 2016-10-12 15:15:18 |    7 |
+-------------+---------------------+---------------------+------+
```

**Comment by shlomi-noach** [comment]
And this leads to an interesting observation: you cannot `--migrate-on-replica` without actually _knowing_ the master's `time_zone`. Reading the master's `time_zone` is essential to solving the math.

**Comment by mignaulo** [comment]
Luckily for me, master, replica and gh-ost are all in the same time zone.

Finding the timezone offset is interesting, however that offset is not constant due to daylight saving. if data was from a different period of the year and had a different offset, the data would not be correct.

I'm not sure this is fixable in every situation short of using `FROM_UNIXTIME(timestamp)` for timestamp fields when doing binlog apply and letting the master figure out the time zone situation. Do you think that is an option?

**Comment by shlomi-noach** [comment]
> however that offset is not constant due to daylight saving. if data was from a different period of the year and had a different offset, the conversion would not be correct.

I disagree, having discussed this with @jonahberquist : once the `DATETIME` data is writter in the table, it loses timezone context. At time of conversion (even normal `ALTER`) MySQL will just take the column value for granted and apply for current timezone. An this is the same action we will be taking.

**Comment by shlomi-noach** [comment]
@mignaulo , can you kindly test the new and shiney https://github.com/github/gh-ost/pull/269 ?

**Comment by shlomi-noach** [comment]
For prosperity, @MarkLeith pointed me to this post on old-vs-new time formats: http://mysqlserverteam.com/how-to-easily-identify-tables-with-temporal-types-in-old-format/

**Comment by mignaulo** [comment]
@shlomi-noach I could not get this build to not do the right thing, regardless of applier or master timezone. Very nice! 👍

**Comment by shlomi-noach** [comment]
Too many (more than one) negatives. I counted an even number therefore concluded the outcome is a positive. Suspicion supported by the word "nice".

**Comment by shlomi-noach** [comment]
closed by #269
